proposals/p1885.md

```diff
@@ -0,0 +1,458 @@
+
+#### Interface and usage
+
+```rust
+// Carbon definition
+interface Consumable() {
+    let ElementType:! Type
+    fn Consume[me: Self]() -> Optional(ElementType);
+}
+
+// Usage
+class MyContainer {
+    ...
+    impl as Consumable where .ElementType = i32 {
+        fn Consume[me: Self]() -> Optional(ElementType) {
+            ...
+        }
+    }
+}
+
+var container: MyContainer = {...};
+for (var a: auto in container) {}
+// `container` is now empty
+```
+
+#### TODO
+
+-   Size method?
+
+### Iterable
+
+The `Iterable` interface represents a generic sequential-access type of
+container. It relies on two new interfaces:
+
+-   `Iteratable`: Interface for a type that can return an iterator and be
+    iterated over
+-   `Iterator`: Interface for an iterator-like object
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   Get an iterator with `GetIterator()`
+-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
+    -   If the `Optional` is null, break the loop. Otherwise return the value
+        held by the Optional.
+
+#### Interface and usage
+
+```rust
+// Basic immutable iterator
+interface Iterator() {
+    let ElementType:! Type;
+    fn Next[me: Self]() -> Optional(ElementType);
+}
+
+// Iterable type
+interface Iterable() {
+    let IteratorType:! Iterator;
+    fn GetIterator[me: Self]() -> IteratorType;
+}
+```
+
+These interfaces can be used as such:
+
+```rust
+class MyContainer() {
+    ...
+    external impl as Iterable {
+        class IteratorType { // Associated type as member type
+            impl as Iterator where .ElementType = i32 {
+                ...
+
+                fn Next[addr me: *Self]() -> Optional(ElementType) {
+                    ...
+                    return if atEnd then .Empty else currItem;
+                }
+            }
+        }
+        fn GetIterator[me: Self]() -> IteratorType {
+            return IteratorType { ... }
+        }
+    }
+}
+```
+
+#### Iterator: `begin()` & `end()`
+
+The solution highlighted embeds the end condition in `Next()`, returns a null
+`Optional` at the last elements. This guarantees that `Iterator`s cannot be used
+improperly (code safety goal), and exposes a simple interface, but does have a
+performance cost (cost of the `Optional`).
+
+If we want to retain familiarity with C++ iterators and eliminate this cost, one
+option would be to expose `Begin()` and `End()` members functions for
+`Iterable`, and allow unsafe access. The lack of pointer arithmetic might hinder
+performance improvements for unsafe use, but other solutions may exist.
+
+#### Mutable, Bidirectional, and others
+
+The current `Iterator` interfaces allows simple extension to add features such
+as bidirectional access (adding `Prev()` method), or mutable access
+(`GetMutable[addr me: Self*]() -> ElementType*`, or `Set()`)
```

Comment:
Should there be a non-mutable `Get` as well? It seems like it would be awkwardly non-orthogonal that if you want to access the element without mutating the iterator, you need to obtain mutable access to the element.

Reply:
Correct. This was meant to be `NextMutable` instead, and doesn't make much sense alone. But based on the interoperability conversation, I will go ahead and add `Get`, which will have the benefit of providing a symmetric solution for this.